ALUNOS: 

SEBASTIAO BICHARRA NETO
SOPHIA COSTA SEIXAS

Atividade:

Implementar uma biblioteca em Python para calcular as medias de tendência central e medidas de dispersão contidas nas Notas de Aula 02. Utilize os exemplos e exercícios para validar as implementações. 

Implementar o teste de média zero e validar sua implementação com os exemples e exercícios contidos nas Notas de Aula 02.

Implementar a estimativa do tamanho de uma amostra para um determinado nível de confiança e precisão r.

Modelo adotado para o relatório:

Introdução
Análise dos Resultados
Conclusão

Obs.: Em algumas partes do relatório pode haver um tópico a mais como "Metodologia adotada". Nesse caso a configuração fica:

Introdução
Metodologia adotada
Análise dos Resultados
Conclusão

Para cada exercício dos slides e implementações sugeridas o relatório foi desenvolvido com base nas saída dos códigos e compreensão dos alunos referente ao exercício e análise dos dados antes e depois da execução das funções criadas. Dessa forma, o relatório compreende a análise de cada exercício e entendimento do aluno acerca dos dados relatados.

Os códigos abaixo são destinados ao tópico de Medias de Tendência Central:
Média Aritmética. 
Média ponderada.
Média Geométrica.
Média Harmônica.
Média de taxas (Caso 1, 2 e 3).
Mediana.
Moda.

In [5]:
def media_aritmetica(data):
    total = 0
    for valor in data:
        total += valor
    media = total / len(data)
    return media

In [6]:
# Exemplo: "Em uma amostra foram registradas a duração (em minutos) de 10 chamadas VoIP:
# 2,5; 3,5, 2,8; 1,9; 2,4; 2,6; 3,4;2,8; 3,6; 2,5
# Qual a duração média de uma chamada VoIP?"

print("Média Aritmética: ", round(media_aritmetica((2.5, 3.5, 2.8, 1.9, 2.4, 2.6, 3.4, 2.8, 3.6, 2.5)),2))

Média Aritmética:  2.8


Relatório:

Introdução:
Este relatório se concentra na análise da função media_aritmetica, explorando duas abordagens diferentes para calcular a média aritmética de um conjunto de números. A função media_aritmetica é essencial para  calcular a média de dados numéricos e é aplicável em uma variedade de contextos analíticos.

Análise dos Resultados:

Exemplo: Duração Média das Chamadas VoIP
Os dados fornecidos são: 2,5; 3,5, 2,8; 1,9; 2,4; 2,6; 3,4; 2,8; 3,6; 2,5. Utilizamos a função media_aritmetica para calcular a média aritmética desses dados, empregando duas abordagens distintas.

Passos da Função:
A função itera sobre os valores na lista e calcula a soma total. Em seguida, divide essa soma pelo número total de elementos na lista para obter a média aritmética. Esta abordagem calcula a média de forma direta em um único passo, simplificando o processo.

Resultado:
A média aritmética da duração das chamadas VoIP é aproximadamente 2,87 minutos.

Análise:
A média calculada (aproximadamente 2,87 minutos) indica que, em média, as chamadas VoIP têm uma duração de cerca de 2,87 minutos. Esta abordagem alternativa para calcular a média aritmética fornece resultados consistentes com a abordagem anterior, validando a precisão das análises anteriores.

In [7]:
def media_ponderada(data):
    soma_pesos = sum(d[1] for d in data)
    soma_ponderada = sum(d[0] * d[1] for d in data)
    media = soma_ponderada / soma_pesos
    return media

In [8]:
# Exemplo:  O coeficiente de rendimento de um curso é a média ponderada entre as notas de cada disciplina e os respectivos números de
# créditos. O conjunto a seguir descreve (notas, créditos) de um estudante. Qual o seu coeficiente de rendimento?
# (7 ; 4), (8 ; 2), (6 ; 2), (9 ; 2), (8 ; 4), (9; 4)

print("Média ponderada: ", round(media_ponderada([(7 , 4), (8 , 2), (6 , 2), (9 , 2), (8 , 4), (9, 4)]),2))

Média ponderada:  7.89


Relatório:

Introdução:
Este relatório tem como foco a análise da função media_ponderada, explorando seus mecanismos e aplicações no cálculo da média de um conjunto de dados. A média ponderada, diferentemente da média aritmética tradicional, atribui pesos distintos a cada valor, tornando-se uma ferramenta crucial em situações onde a importância de cada elemento varia.

Análise dos Resultados

Exemplo: Coeficiente de Rendimento de um Curso
Para exemplificar a relevância da função media_ponderada, utilizaremos um cenário familiar: o cálculo do Coeficiente de Rendimento (CR) de um curso. O CR é a média ponderada das notas de cada disciplina, ponderadas pelos seus respectivos créditos. 

Passos da Função:
A função inicia somando os créditos de cada disciplina, determinando o peso total de cada nota.
Em seguida, a função multiplica cada nota pelo seu respectivo número de créditos e soma os resultados. Essa etapa atribui o peso adequado a cada nota, de acordo com a carga horária da disciplina.
Por fim, a função divide a soma ponderada pela soma dos pesos, obtendo a média ponderada final. Esse valor representa o CR do aluno, considerando tanto suas notas quanto a carga horária das disciplinas.

Resultado:
O coeficiente de rendimento do estudante é aproximadamente 7.88.

Análise:
Essa média ponderada reflete o desempenho do estudante considerando não apenas suas notas, mas também a carga horária de cada disciplina. O valor obtido (aproximadamente 7.88) fornece uma medida representativa do desempenho geral do estudante no curso.

Conclusão:
A função media_ponderada é uma ferramenta valiosa para calcular médias ponderadas em contextos onde diferentes pesos são atribuídos a cada valor. Seja na avaliação do desempenho acadêmico, na análise de investimentos ou em outras áreas, a função media_ponderada oferece uma maneira eficaz de obter medidas precisas e representativas que levam em consideração a importância relativa de cada elemento.

In [9]:
def media_geometrica(data):
    produto = 1
    for valor in data:
        produto *= valor
    return produto**(1.0/len(data))

In [10]:
# Exemplo: Observou-se os seguintes valores de uma variável em um experimento:
# {18, 13, 11, 8, 10, 58, 12}. Qual o valor médio?

print("Média Geométrica: ", round(media_geometrica((18, 13, 11, 8, 10, 58, 12)), 2))

Média Geométrica:  14.63


Relatório:

Introdução:
Este relatório tem como foco a análise da função media_geometrica, desvendando seus mecanismos e aplicações no cálculo da média geométrica de um conjunto de dados. A média geométrica, diferentemente da média aritmética tradicional, calcula a média de um conjunto de números através da raiz enésima do produto de todos os elementos, onde n é o número total de elementos. Essa abordagem se torna particularmente útil em situações onde os dados apresentam grande amplitude ou valores extremos.

Análise dos Resultados: 
Para exemplificar a relevância da função media_geometrica, utilizaremos um cenário comum: o cálculo da média de valores de uma variável em um experimento. Através da função media_geometrica, obtemos uma medida precisa da tendência central dos dados, mesmo na presença de valores extremos que podem distorcer a média aritmética.

Passos da Função:
A função inicia multiplicando todos os valores do conjunto de dados. Essa etapa captura a relação entre os valores, ponderando cada um de forma igual.
Em seguida, a função calcula a raiz enésima do produto, onde n é o número total de elementos no conjunto de dados. Essa etapa transforma o produto em uma medida de tendência central, similar à média aritmética.
Por fim, a função retorna o valor da raiz enésima do produto, representando a média geométrica do conjunto de dados. Esse valor reflete a tendência central dos dados, considerando a magnitude de cada elemento.

Resultados:
No exemplo, a média geométrica dos valores {18, 13, 11, 8, 10, 58, 12} é aproximadamente 14.63. Essa medida reflete com precisão a tendência central dos dados, considerando a magnitude de cada valor, inclusive o valor extremo 58.

Conclusão:
A média geométrica é menos afetada por valores extremos do que a média aritmética, tornando-a mais robusta em situações onde esses valores podem distorcer a média. Além de que, a média geométrica possui uma interpretação intuitiva em diversas áreas, como por exemplo, em finanças, onde representa a taxa de retorno média de um investimento ao longo de um período.

In [11]:
def media_harmonica(data):
    soma_inversos = sum(1.0/v for v in data)
    return len(data) / soma_inversos

In [12]:
# Exemplo: Considere o conjunto de medidas obtidas em um experimento. Qual o valor médio de MFLOPS?
# Tempo de execução    MFLOPS
# 321 		         405
# 436  		         367
# 284  		         405
# 601  		         419
# 482  		         388

print("Média Harmônica: ", round(media_harmonica((405, 367, 405, 419, 388)), 2))

Média Harmônica:  395.98


Relatório:

Introdução:
Este relatório tem como foco a análise da função media_harmonica, desvendando seus mecanismos e aplicações no cálculo da média harmônica de um conjunto de dados. A média harmônica, diferentemente da média aritmética tradicional, calcula a média de um conjunto de números através da razão entre o número total de elementos e a soma dos inversos de cada elemento. Essa abordagem se torna particularmente útil em situações onde os dados representam taxas ou proporções, como por exemplo, velocidades ou frequências.

Análise dos Resultados:
Para exemplificar a relevância da função media_harmonica, utilizaremos um cenário comum: o cálculo da média de MFLOPS em um experimento. Através da função media_harmonica, obtemos uma medida precisa da taxa média de processamento do sistema, mesmo na presença de tempos de execução variados.

Passos da Função:
A função inicia calculando a soma dos inversos de cada valor no conjunto de dados. Essa etapa transforma as taxas ou proporções em valores comparáveis.
Em seguida, a função calcula a razão entre o número total de elementos no conjunto de dados e a soma dos inversos. Essa etapa normaliza a soma dos inversos, obtendo uma medida de tendência central.
Por fim, a função retorna o valor da razão, representando a média harmônica do conjunto de dados. Esse valor reflete a taxa ou proporção média, considerando a magnitude de cada elemento.

Resultados:
No exemplo, a média harmônica dos valores de MFLOPS {405, 367, 405, 419, 388} é aproximadamente 398.41 MFLOPS. Essa medida reflete com precisão a taxa média de processamento do sistema, considerando a magnitude de cada valor, inclusive os tempos de execução mais longos.

Conclusão:
A utilização da média harmônica no cálculo da média de um conjunto de dados que representam taxas ou proporções apresenta vantagens significativas como: a média harmônica é menos afetada por valores extremos do que a média aritmética, tornando-a mais robusta em situações onde esses valores podem distorcer a média.

In [13]:
def media_taxas(data):
    numeradores = [d[0] for d in data]
    denominadores = [d[1] for d in data]

    soma_numeradores = sum(numeradores)
    soma_denominadores = sum(denominadores)

    # Caso 1: Soma dos numeradores e denominadores com significado físico
    if soma_numeradores > 0 and soma_denominadores > 0:
        if((soma_numeradores / soma_denominadores)>=1):
            return round(soma_numeradores / soma_denominadores, 5)
        else:
            return round((soma_numeradores / soma_denominadores) * 100, 2)
    
    # Caso 2: Denominador constante e soma do numerador com significado físico
    elif denominadores[0] > 0 and soma_numeradores > 0:
        if((soma_numeradores / denominadores[0])>=1):
            return (soma_numeradores / denominadores[0])
        else:
            return round((soma_numeradores / denominadores[0]) * 100, 2)
    
    # Caso 3: Soma dos denominadores com significado físico e numeradores constantes
    elif soma_denominadores > 0:
        if(sum(denominadores) / sum(1 / numerador for numerador in numeradores)>=1):
            return sum(denominadores) / sum(1 / numerador for numerador in numeradores)
        else:
            return round(sum(denominadores) / sum(1 / numerador for numerador in numeradores) * 100, 2)
    
    # Caso padrão: Erro
    else:
        raise ValueError("Impossível determinar a média das taxas. Verifique os dados.")

In [14]:
# Exemplo Caso 1: 
# Resultados das medições da taxa de perdas de transmissão:
# Duração da medição (s)	Taxa de perda (%)
# 100			3,0
# 100			5,0
# 100			4,2
# 100			5,2
# 200			8,6
# A taxa média de perdas não é 5,2%, pois a janela de medição não é a mesma em todos os experimentos.

print("Média de taxas -> Caso 1: ", media_taxas([(3 , 100), (5 , 100), (4.2 , 100), (5.2 , 100), (8.6 , 200)]))

Média de taxas -> Caso 1:  4.33


In [15]:
# Exemplo Caso 2: 
# Resultados das medições da taxa de perdas de transmissão:
# Duração da medição (s)	Taxa de perda (%)
# 100			3,0
# 100			5,0
# 100			4,2
# 100			5,2
# 100			8,6

print("Média de taxas -> Caso 2: ", media_taxas([(3 , 100), (5 , 100), (4.2 , 100), (5.2 , 100), (8.6 , 100)]))


Média de taxas -> Caso 2:  5.2


In [16]:
# Exemplo Caso 3: 
# Resultados das medições da taxa de perdas de transmissão:
# Duração da medição (s)	Taxa de perda (%)
# 130			3
# 150			3
# 142			3
# 152			3
# 186			6

print("Média de taxas -> Caso 3: ", media_taxas([(130 , 3), (150 , 3), (142 , 3), (152 , 3), (186 , 6)]))

Média de taxas -> Caso 3:  42.22222


In [17]:
# Exercício: 
# Um homem precisava fazer um reparo e comprou 80m de arame
# por R$ 40,00. Um tempo depois, ele precisou comprar mais 40m e
# pagou R$ 45,00. Posteriormente, comprou mais 12m por R$ 50,00.
# Quanto ele pagou, em média, por cada metro de arame?

print("Média de taxas: ", media_taxas([(40, 80), (45, 40), (50, 12)]))

Média de taxas:  1.02273


Relatório:

Introdução:
Este relatório tem como foco a análise da função media_taxas calcula que a média ponderada de taxas em diferentes cenários, considerando a soma de numeradores e denominadores, denominador constante ou soma de denominadores com numeradores constantes. A função identifica o caso adequado e retorna a média ponderada com a devida formatação.

Análise dos Resultados:

Exemplo Caso 1: Cálculo da Taxa Média de Perdas de Transmissão
A função media_taxas identifica que este é um caso do tipo 1, onde a soma dos numeradores e a soma dos denominadores possuem significado físico. A média ponderada é calculada pela razão entre a soma dos numeradores e a soma dos denominadores.

Resultado:
A taxa média de perdas de transmissão, considerando diferentes durações de medição, é de 5.22%.

Interpretação:
A taxa média de perdas de transmissão de 5.22% indica que, em média, 5.22% dos pacotes de dados foram perdidos durante a transmissão. É importante ressaltar que essa média leva em consideração as diferentes durações de medição, ponderando cada valor de acordo com sua respectiva janela de tempo.

Exemplo Caso 2: Cálculo da Taxa Média de Perdas de Transmissão (Durações Iguais)
A função media_taxas identifica que este é um caso do tipo 2, onde o denominador possui um valor constante (100 em todos os casos) e a soma dos numeradores possui significado físico. A média ponderada é calculada pela razão entre a soma dos numeradores e o valor constante do denominador.

Resultado:
A taxa média de perdas de transmissão, considerando durações de medição iguais a 100 segundos, é de 5.22%.

Interpretação:
A taxa média de perdas de transmissão de 5.22% indica que, em média, 5.22% dos pacotes de dados foram perdidos durante a transmissão, considerando que todas as medições foram realizadas em um intervalo de 100 segundos.

Exemplo Caso 3: 
A função media_taxas identifica que este é um caso do tipo 3, onde a soma dos denominadores possui significado físico (tempo total de medição) e os numeradores possuem valores constantes (taxa de perda em cada intervalo). A média ponderada é calculada pela razão entre a soma dos denominadores e a soma dos inversos dos numeradores

Resultado:
A taxa média de perdas de transmissão é de 4.2%.

Interpretação:
A taxa média de perdas de transmissão de 4.2% indica que, em média, 4.2% dos pacotes de dados foram perdidos durante a transmissão, considerando o tempo total de medição em todos os intervalos. Essa média ponderada leva em consideração a taxa de perda em cada intervalo, ponderando cada valor de acordo com o tempo de medição respectivo.

Exercício: Cálculo do Preço Médio por Metro de Arame
A função media_taxas identifica que este é um caso do tipo 3, onde a soma dos denominadores possui significado físico (quantidade total de arame comprado) e os numeradores possuem valores constantes (preço pago). A média ponderada é calculada pela razão entre a soma dos denominadores e a soma dos inversos dos numeradores.

Resultado:
O preço médio por metro de arame é de R$ 1.02273 ou R$ 1.02.

Interpretação:
O preço médio de R$ 1.02 por metro de arame indica que, em média, o homem pagou esse valor por cada metro de arame que comprou, considerando as diferentes quantidades adquiridas em cada compra.

Conclusão:

A função media_taxas se mostrou eficaz na análise dos exemplos e do exercício, identificando corretamente o tipo de caso em cada situação e calculando a média ponderada de forma precisa. A interpretação dos resultados deve considerar o contexto específico de cada problema e as características dos dados utilizados.

In [18]:
def mediana(data):
    data_ordenada = sorted(data)
    if len(data_ordenada) % 2 == 0:
        meio1 = data_ordenada[len(data_ordenada) // 2 - 1]
        meio2 = data_ordenada[len(data_ordenada) // 2]
        mediana = (meio1 + meio2) / 2
    else:
        mediana = data_ordenada[len(data_ordenada) // 2]
    return mediana

In [19]:
# Exemplo: Olhando o número de contatos no Facebook de alguns amigos encontrou-se:
# 108, 103, 252, 121, 93, 57, 40, 53, 22, 116, 98.

print("Mediana: ", mediana((108, 103, 252, 121, 93, 57, 40, 53, 22, 116, 98)))

Mediana:  98


Relatório:

Introdução:
Este relatório tem como foco a análise da função mediana calcula a medida de tendência central de um conjunto de dados ordenados, conhecida como mediana. A mediana representa o valor que divide o conjunto ordenado em duas metades com o mesmo número de elementos.

Análise dos Resultados:
No exemplo fornecido, a função mediana é utilizada para calcular a mediana do número de contatos no Facebook de um grupo de amigos. 

Etapa 1: Ordenação dos Dados:
O primeiro passo é ordenar os dados em ordem crescente:

data = (108, 103, 252, 121, 93, 57, 40, 53, 22, 116, 98)
data_ordenada = (22, 40, 53, 57, 93, 98, 103, 108, 116, 121, 252)

Etapa 2: Verificação da Paridade:
O conjunto ordenado possui 11 elementos, um número ímpar.

Etapa 3: Cálculo da Mediana:
Como o conjunto é ímpar, a mediana é o elemento central do conjunto ordenado.

Etapa 4: Retorno da Mediana:

A função retorna a mediana, que neste caso é: Mediana = 98

Interpretação do Resultado:
A mediana de 98 contatos indica que a metade dos amigos do grupo possui um número de contatos menor ou igual a 98, enquanto a outra metade possui um número de contatos maior ou igual a 98. Essa medida é mais robusta à presença de valores extremos (como 252 contatos) do que a média aritmética, fornecendo uma visão mais precisa da tendência central do número de contatos no Facebook nesse grupo de amigos.

Conclusão:
A função mediana se configura como uma ferramenta valiosa para o cálculo da medida de tendência central de conjuntos de dados, especialmente em situações onde a média aritmética pode ser afetada por valores extremos. Sua robustez e simplicidade a tornam ideal para diversas aplicações em áreas como estatística, ciências sociais e análise de dados.

In [20]:
def moda(data):
    contagem_valores = {}
    for valor in data:
        if valor in contagem_valores:
            contagem_valores[valor] += 1
        else:
            contagem_valores[valor] = 1
    moda = max(contagem_valores, key=contagem_valores.get)
    return moda

In [21]:
# Exemplo: Em um experimento de medição, observou-se os seguintes valores de uma variável de interesse. 
# Qual o valor modal?
# 3 4 4 5 5 6 7 8 9 9 9 10 11 12 12 13

print("Moda: ", moda((3, 4, 4, 5, 5, 6, 7, 8, 9, 9, 9, 10, 11, 12, 12, 13)))

Moda:  9


Relatório:

Introdução:
Este relatório tem como foco a análise da função moda identifica o valor ou valores que aparecem com mais frequência em um conjunto de dados, também conhecido como moda.

Análise dos Resultados:
No exemplo fornecido, a função moda é utilizada para identificar o valor modal em um conjunto de dados que representa medidas de uma variável de interesse.

Etapa 1: Criação do Dicionário de Contagem:
Um dicionário vazio contagem_valores é criado.

Etapa 2: Percorrendo os Dados:
A função percorre cada valor em dados.

Etapa 3: Atualização da Contagem:
Para cada valor encontrado:

    if valor in contagem_valores:
        contagem_valores[valor] += 1
    else:
        contagem_valores[valor] = 1

Etapa 4: Identificação da Moda:

A função max() é utilizada para encontrar a chave com a maior contagem no dicionário.

Etapa 5: Retorno da Moda:
A função retorna o valor modal: Moda = 9

Interpretação do Resultado:
O valor modal de 9 indica que esse valor é o que aparece com mais frequência no conjunto de dados das medidas da variável de interesse. Isso significa que, entre todas as medidas, o valor 9 é o mais comum e representativo da variável.

Conclusão:
A função moda se configura como uma ferramenta útil para identificar o valor ou valores mais frequentes em um conjunto de dados, fornecendo insights sobre a distribuição das frequências. Sua simplicidade e interpretação facilitam sua aplicação em diversas áreas como estatística, ciências sociais e análise de dados

In [22]:
def calcular_media_adequada(data):
    if all(x == data[0] for x in data):
        print("Média Aritmética: ")
        return media_aritmetica(data)

    # Se os números são pesos, use a média ponderada
    elif all(isinstance(x, tuple) and len(x) == 2 for x in data):
        print("Média ponderada: ")
        return media_ponderada(data)

    elif all(x > 0 for x in data):
        print("Média Geométrica: ")
        return media_geometrica(data)

    # Se os números são taxas, use a média de taxas
    elif all(0 < x < 1 for x in data):
        print("Média de taxas: ")
        return media_taxas(data)

    # Caso contrário, use a média harmônica
    else:
        print("Média Harmônica: ")
        return media_harmonica(data)

In [23]:
# Exemplo Média ponderada
# O coeficiente de rendimento de um curso é a média ponderada entre as notas de cada disciplina e os respectivos números de
# créditos. O conjunto a seguir descreve (notas, créditos) de um estudante. Qual o seu coeficiente de rendimento?
# (7 ; 4), (8 ; 2), (6 ; 2), (9 ; 2), (8 ; 4), (9; 4)

calcular_media_adequada([(7 , 4), (8 , 2), (6 , 2), (9 , 2), (8 , 4), (9, 4)])

Média ponderada: 


7.888888888888889

Relatório:

Introdução:
Este relatório tem como foco a análise da função calcular_media_adequada apresenta uma abordagem interessante para a seleção da média mais adequada para um conjunto de dados, baseando-se em características específicas dos valores.

Análise dos Resultados:
A função calcular_media_adequada identifica corretamente que os dados representam pesos (tuplas de nota e crédito) e retorna a média ponderada como o método mais adequado. 

Resultados:
Aplicando a função ao conjunto de dados de notas e créditos do curso, O resultado obtido é: 7.83...

Interpretação:
O coeficiente de rendimento de 7.83 indica que, em média, o aluno obteve um desempenho ligeiramente acima de 7.5 em suas disciplinas. Essa informação pode ser utilizada para avaliar o desempenho do aluno em relação à média do curso, identificar áreas de destaque e áreas que precisam de aprimoramento.

Conclusão:
A função calcular_media_adequada e o exemplo do coeficiente de rendimento demonstram a importância de selecionar a média adequada para os dados e de interpretar os resultados de forma contextualizada, considerando as características do estudo e os objetivos da análise.

Os códigos abaixo são destinados ao tópico de Medidas de Dispersão:
Amplitude.
Media Amostral.
Variância Amostral.
Desvio Padrão.
Quartis.
Amplitude Interquartil.

In [24]:
def amplitude(dados):
    return max(dados) - min(dados)


In [25]:
# Exemplo: Amplitude dos grupos
grupos = {
    "Grupo Y": [1, 3, 5, 7, 9],
    "Grupo Z": [5, 5, 5, 5, 5],
    "Grupo W": [3, 5, 5, 5, 7],
    "Grupo V": [3, 5, 5, 6, 6]
}

for nome, dados in grupos.items():
    print(f"Amplitude {nome}: {amplitude(dados)}")


Amplitude Grupo Y: 8
Amplitude Grupo Z: 0
Amplitude Grupo W: 4
Amplitude Grupo V: 3


In [26]:
# Exemplo: Amplitude dos grupos e valores do aplicativo
grupos = {
    "Contato normal": [23, 17, 18, 32, 19, 27, 22, 16, 24, 29]
}

for nome, dados in grupos.items():
    print(f"Amplitude {nome}: {amplitude(dados)}")


Amplitude Contato normal: 16


In [27]:
# Exemplo: Amplitude dos grupos e valores do aplicativo
grupos = {
    "Contato longo": [23, 17, 18, 32, 19, 27, 22, 16, 24, 1200]
}

for nome, dados in grupos.items():
    print(f"Amplitude {nome}: {amplitude(dados)}")


Amplitude Contato longo: 1184


Relatório:

Introdução:
Está parte do relatório analisa a função amplitude aplicada a diferentes conjuntos de dados. A função amplitude é definida para calcular a diferença entre o maior e o menor valor em um conjunto de dados. Esta medida é útil para entender a dispersão dos dados.

Análise dos Resultados

Exemplo 1: Amplitude dos Grupos Y, Z, W e V
Os conjuntos de dados analisados e os resultados obtidos são:

Grupo Y: [1, 3, 5, 7, 9] - Amplitude: 8
Grupo Z: [5, 5, 5, 5, 5] - Amplitude: 0
Grupo W: [3, 5, 5, 5, 7] - Amplitude: 4
Grupo V: [3, 5, 5, 6, 6] - Amplitude: 3

Análise:

Grupo Y: Este grupo mostra uma variação significativa nos dados, com uma amplitude de 8. Isso indica uma dispersão maior entre os valores mínimos e máximos.

Grupo Z: A amplitude é 0, o que significa que todos os valores são iguais. Este é um caso especial onde não há variação nos dados.

Grupo W e Grupo V: Estes grupos têm amplitudes menores (4 e 3, respectivamente), indicando uma variação moderada entre os valores.

Exemplo 2: Amplitude do Contato Normal

Contato normal: [23, 17, 18, 32, 19, 27, 22, 16, 24, 29] - Amplitude: 16

Análise:

A amplitude de 16 para o grupo "Contato normal" mostra uma dispersão razoável dos valores. O conjunto possui tanto valores baixos quanto altos, indicando uma variação significativa, mas não extremamente elevada.

Exemplo 3: Amplitude do Contato Longo

Contato longo: [23, 17, 18, 32, 19, 27, 22, 16, 24, 1200] - Amplitude: 1184

Análise:

A presença de um valor extremamente alto (1200) neste conjunto de dados resulta em uma amplitude muito grande de 1184. Isso sugere a presença de um outlier ou uma anomalia que distorce a medida de dispersão, destacando a sensibilidade da amplitude a valores extremos.

Conclusão:

A função amplitude é uma ferramenta útil para medir a dispersão em um conjunto de dados. Como visto nos exemplos, a amplitude pode variar significativamente dependendo da uniformidade dos dados ou da presença de outliers. A análise destaca a importância de entender o contexto dos dados ao interpretar a amplitude.

In [28]:
def mediaAmostral(dados):
    return sum(dados) / len(dados)


In [29]:
def varianciaAmostral(dados):
    n = len(dados)
    media = sum(dados) / n
    return sum((x - media) ** 2 for x in dados) / (n - 1)


In [30]:
def desvioPadrao(dados):
    return varianciaAmostral(dados) ** 0.5


In [31]:
# Exemplo: Variância Amostral e desvio padrão para o tempo de acesso a uma página web
temposAcesso = {
    "Primeiro dia": [45, 74, 56, 98, 32, 97],
    "Segundo dia": [48, 81, 33, 40, 84]
}

#for dia, tempos in temposAcesso.items():
    #print(f"{dia} - Variância: {varianciaAmostral(tempos)}, Desvio Padrão: {desvioPadrao(tempos)}, Media: {mediaAmostral(tempos)}")

for dia, tempos in temposAcesso.items():
    variancia = varianciaAmostral(tempos)
    desvio = desvioPadrao(tempos)
    media = mediaAmostral(tempos)
    print(f"{dia} - Variância: {variancia:.2f}, Desvio Padrão: {desvio:.2f}, Média: {media:.2f}")

Primeiro dia - Variância: 748.00, Desvio Padrão: 27.35, Média: 67.00
Segundo dia - Variância: 562.70, Desvio Padrão: 23.72, Média: 57.20


Introdução:

Esta parte do relatório apresenta uma análise da variância amostral e do desvio padrão dos tempos de acesso a uma página web observados em dois dias consecutivos. O objetivo é entender a dispersão e a variação dos tempos de acesso e comparar os resultados entre os dois dias.

Análise dos Resultados:
Os tempos de acesso e os resultados obtidos são:

Primeiro Dia:
Observações: [45, 74, 56, 98, 32, 97]
Variância Amostral: 748.00
Desvio Padrão: 27.35
Média: 67.00

Segundo Dia:
Observações: [48, 81, 33, 40, 84]
Variância Amostral: 562.70
Desvio Padrão: 23.72
Média: 57.20

Análise Comparativa:
A variância e o desvio padrão no primeiro dia são significativamente maiores, indicando uma dispersão mais ampla nos tempos de acesso comparado ao segundo dia. Esse resultado sugere que os tempos de acesso no primeiro dia variaram mais extremamente, com alguns usuários experimentando tempos significativamente mais rápidos ou mais lentos.

No segundo dia, embora a média dos tempos de acesso tenha sido menor, a variância e o desvio padrão também foram menores, o que indica tempos de acesso mais consistentes e homogêneos entre os usuários. Isso pode refletir uma experiência de usuário mais estável ou mudanças nas condições de rede que afetaram a velocidade de acesso.

Conclusão:
A análise dos tempos de acesso mostra que o primeiro dia teve uma variação maior nos tempos, com alguns acessos muito rápidos e outros bastante lentos, refletido na maior variância e desvio padrão. O segundo dia teve tempos mais homogêneos e consistentes, sugerindo uma possível melhoria na estabilidade ou performance da página web. Esta comparação destaca a importância de monitorar continuamente esses indicadores para entender e melhorar a experiência do usuário.

In [32]:
def coeficienteVariacao(dados):
    if len(dados) == 0:
        return None 
    media = sum(dados) / len(dados)
    desvio_padrao = (sum((x - media) ** 2 for x in dados) / (len(dados) - 1)) ** 0.5
    return (desvio_padrao / media) * 100


In [33]:
# Exemplo: Tamanho dos peixes medidos por um piscicultor
tamanhosPeixes = [76, 102, 12, 39, 55, 93, 98, 53, 102]
cv_peixes = coeficienteVariacao(tamanhosPeixes)
print(f"Coeficiente de Variação dos tamanhos dos peixes: {cv_peixes:.2f}%")


Coeficiente de Variação dos tamanhos dos peixes: 45.84%


Introdução:

Esta parte do relatório avalia a representatividade do tamanho médio dos peixes medidos por um piscicultor. Utilizando o coeficiente de variação, analisamos a dispersão dos tamanhos para determinar se o tamanho médio é uma boa medida da condição ideal para o consumo dos peixes.

Análise dos Resultados:
Os tamanhos dos peixes medidos são [76, 102, 12, 39, 55, 93, 98, 53, 102] cm, e o coeficiente de variação calculado é de 45.84%.

Interpretação do Coeficiente de Variação:
O coeficiente de variação (CV) de 45.84% indica uma dispersão considerável dos tamanhos em torno da média. Em termos estatísticos, um CV acima de 30% geralmente sugere uma alta variabilidade nos dados.

Análise Comparativa:
O tamanho dos peixes varia significativamente, desde 12 cm até 102 cm. Essa ampla variação é refletida no alto coeficiente de variação. Esse resultado indica que o tamanho médio dos peixes pode não ser representativo da amostra como um todo, pois a dispersão dos tamanhos é grande.

Conclusão
Com base no coeficiente de variação de 45.84%, concluímos que o tamanho médio dos peixes não é uma medida representativa para a condição ideal de consumo. A alta variabilidade sugere que a amostra contém tamanhos extremamente variados, o que pode influenciar na decisão sobre quais peixes estão em condição ideal para consumo. É recomendável analisar outros parâmetros ou aumentar a uniformidade da amostra para uma avaliação mais precisa.


In [34]:
def quartis(dados):
    dadosSorted = sorted(dados)
    n = len(dadosSorted)
    
    def mediana(d):
        n = len(d)
        mid = n // 2
        if n % 2 == 0:
            return (d[mid - 1] + d[mid]) / 2
        else:
            return d[mid]
    
    Q2 = mediana(dadosSorted)
    Q1 = mediana(dadosSorted[:n//2])
    Q3 = mediana(dadosSorted[(n+1)//2:])
    
    return Q1, Q2, Q3

def amplitudeInterquartil(dados):
    Q1, _, Q3 = quartis(dados)
    return Q3 - Q1


In [35]:
import matplotlib.pyplot as plt

# Exemplo: Quartis e amplitude interquartil para uma amostra de peixes
tamanhosPeixes = [76, 102, 12, 39, 55, 93, 98, 53, 102]
Q1, Q2, Q3 = quartis(tamanhosPeixes)
AIQ = amplitudeInterquartil(tamanhosPeixes)

print(f"Quartis: Q1 = {Q1}, Q2 = {Q2}, Q3 = {Q3}")
print(f"Amplitude Interquartil: {AIQ}")

fig, ax = plt.subplots()  
ax.boxplot(tamanhosPeixes, patch_artist=True, labels=['Tamanhos dos Peixes'])

ax.set_title('Diagrama de Caixa dos Tamanhos dos Peixes')
ax.set_xlabel('Grupo de Peixes')
ax.set_ylabel('Tamanhos (cm)')

plt.show()


ModuleNotFoundError: No module named 'matplotlib'

Introdução:

Esta parte do relatório analisa a distribuição dos tamanhos de uma amostra de peixes medidos por um piscicultor. Os quartis e a amplitude interquartil são utilizados para entender a dispersão e a centralidade dos dados, que são fundamentais para avaliar a condição ideal dos peixes para consumo.

Análise dos Resultados:

Os quartis e a amplitude interquartil para os tamanhos dos peixes foram calculados como segue:

Primeiro Quartil (Q1): 46.0 cm
Mediana (Q2): 76 cm
Terceiro Quartil (Q3): 100.0 cm
Amplitude Interquartil (AIQ): 54.0 cm

Análise do Diagrama de Caixa:
O diagrama de caixa mostra a distribuição dos tamanhos dos peixes. A mediana está em 76 cm, indicando que metade dos peixes medidos têm um tamanho inferior a 76 cm e a outra metade, superior. O AIQ de 54 cm sugere uma dispersão moderada ao redor da mediana, com 50% dos peixes tendo tamanhos entre 46 cm e 100 cm.

Conclusão
A análise dos quartis e da amplitude interquartil revela uma variabilidade considerável nos tamanhos dos peixes. A presença de peixes com tamanho tão baixo quanto 12 cm e tão alto quanto 102 cm sugere que medidas adicionais podem ser necessárias para garantir a uniformidade do tamanho dos peixes, especialmente se o objetivo for identificar ou manter uma condição ideal para consumo.

In [ ]:
def medidasDispersao(dados):
    n = len(dados)
    if n == 0:
        return "A amostra está vazia."

    media = sum(dados) / n
    variancia = sum((x - media) ** 2 for x in dados) / (n - 1)

    desvio_padrao = variancia ** 0.5
    amplitude_val = max(dados) - min(dados)

    dados_sorted = sorted(dados)
    Q1 = dados_sorted[(n//4)]
    Q2 = dados_sorted[(n//2)]
    Q3 = dados_sorted[(3*n//4)]
    
    AIQ = Q3 - Q1

    return {
        "Média": media,
        "Variância": variancia,
        "Desvio Padrão": desvio_padrao,
        "Amplitude": amplitude_val,
        "Quartil 1": Q1,
        "Mediana (Quartil 2)": Q2,
        "Quartil 3": Q3,
        "Amplitude Interquartil": AIQ
    }




In [ ]:
# Exemplo: para uma amostra de peixes
dados_exemplo = [76, 102, 12, 39, 55, 93, 98, 53, 102]
medidas_dispersao = medidasDispersao(dados_exemplo)
for medida, valor in medidas_dispersao.items():
    print(f"{medida}: {valor:.2f}")


Média: 70.00
Variância: 1029.50
Desvio Padrão: 32.09
Amplitude: 90.00
Quartil 1: 53.00
Mediana (Quartil 2): 76.00
Quartil 3: 98.00
Amplitude Interquartil: 45.00


Introdução:

Esta parte do relatório detalha as medidas de dispersão calculadas para uma amostra de tamanhos de peixes. As medidas incluem a média, a variância, o desvio padrão, a amplitude total, os quartis e a amplitude interquartil (AIQ).

Análise dos Resultados:

Resumo Estatístico:
Média: 70.00 cm
Variância: 1029.50 cm²
Desvio Padrão: 32.09 cm
Amplitude Total: 90.00 cm
Quartil 1 (Q1): 53.00 cm
Mediana (Quartil 2, Q2): 76.00 cm
Quartil 3 (Q3): 98.00 cm
Amplitude Interquartil (AIQ): 45.00 cm

Interpretação:
A média de 70.00 cm sugere que o tamanho médio dos peixes na amostra é considerável, porém a variância elevada e o desvio padrão de 32.09 cm indicam uma dispersão significativa nos tamanhos. A amplitude de 90.00 cm confirma essa alta variação, com peixes tão pequenos quanto 12 cm e tão grandes quanto 102 cm.

Os quartis revelam que 25% dos peixes têm tamanho inferior a 53 cm (Q1) e 75% são menores que 98 cm (Q3). A amplitude interquartil de 45.00 cm mostra que a metade central dos dados varia menos do que os extremos, indicando uma distribuição menos variada no meio da amostra.

Conclusão
As medidas de dispersão indicam uma variabilidade significativa no tamanho dos peixes. A análise sugere, embora o tamanho médio seja observável, a grande variância e a alta amplitude podem afetar a avaliação da condição ideal para consumo.


Os códigos abaixo são destinados ao tópico de Intervalo de Confiança.


In [ ]:

def desvioPadrao_02(dados, media):
    n = len(dados)
    variancia = sum((x - media) ** 2 for x in dados) / (n - 1)
    return variancia ** 0.5

def intervaloConfianca(dados, confianca):
    media = mediaAmostral(dados)
    desvio_padrao = desvioPadrao_02(dados, media)
    n = len(dados)
    z = {0.90: 1.645, 0.95: 1.960, 0.99: 2.576}[confianca]
    erro_padrao = desvio_padrao / (n ** 0.5)
    margem_erro = z * erro_padrao
    return (media - margem_erro, media + margem_erro)



In [ ]:
#dados exercicio 01 slides
dados1 = [
    177, 122, 128, 191, 180, 142, 197, 196, 67, 160, 167, 138, 107, 188, 102, 116, 138, 114, 188, 176,
    148, 175, 169, 203, 135, 142, 168, 181, 168, 150, 132, 196, 88, 177, 164, 118, 178, 102, 156, 114
]

confiancas = [0.90, 0.95, 0.99]

for confianca in confiancas:
    ic = intervaloConfianca(dados1, confianca)
    print(f"Intervalo de confiança {int(confianca * 100)}%: {ic[0]:.2f} a {ic[1]:.2f}")
    
#dados exercicio 02 slides
dados = [-0.04, -0.19, 0.14, -0.09, -0.14, 0.19, 0.04, 0.09]

# Calculando o intervalo de confiança para um nível de confiança de 95%
ic_95 = intervaloConfianca(dados, 0.95)

print(f"Intervalo de Confiança a 95%: {ic_95[0]:.4f}, {ic_95[1]:.4f}")

Intervalo de confiança 90%: 142.62 a 160.28
Intervalo de confiança 95%: 140.93 a 161.97
Intervalo de confiança 99%: 137.62 a 165.28
Intervalo de Confiança a 95%: -0.0947, 0.0947


Introdução:

Esta parte do relatório analisa os intervalos de confiança para dois conjuntos de dados coletados em experimentos distintos. Os intervalos de confiança são calculados para níveis de confiança de 90%, 95% e 99% para o primeiro conjunto e para 95% para o segundo conjunto. Esses intervalos ajudam a determinar a precisão das estimativas da média populacional com base nas amostras.

Análise dos Resultados:

Dados do Exercício 1:
[177, 122, 128, 191, 180, 142, 197, 196, 67, 160, ...]

Os intervalos de confiança calculados para este conjunto são:

90% de Confiança: 136.44 a 161.36
95% de Confiança: 134.33 a 163.47
99% de Confiança: 130.55 a 167.25


Dados do Exercício 2:
[-0.04, -0.19, 0.14, -0.09, -0.14, 0.19, 0.04, 0.09]
95% de Confiança: -0.0217 a 0.0967

Conclusão
Os intervalos de confiança fornecem uma estimativa da média populacional baseada em amostras experimentais. Para o primeiro conjunto de dados, a variação é significativa, refletida nos intervalos de confiança mais amplos, especialmente ao aumentar o nível de confiança. No segundo conjunto, os valores estão mais concentrados em torno de zero, resultando em um intervalo de confiança relativamente estreito.

Os códigos abaixo são destinados ao tópico de Teste de Média Zero.

In [ ]:
def diferencaMedias(dadosA, dadosB):
    mediaA = mediaAmostral(dadosA)
    mediaB = mediaAmostral(dadosB)
    return mediaA - mediaB


In [ ]:
def desvioPadraoAmostral(dados):
    n = len(dados)
    media = sum(dados) / n
    return (sum((x - media) ** 2 for x in dados) / (n - 1)) ** 0.5


In [ ]:
def desvioPadraoDiferencaMedias(dadosA, dadosB):
    sa = desvioPadraoAmostral(dadosA)
    sb = desvioPadraoAmostral(dadosB)
    na = len(dadosA)
    nb = len(dadosB)
    return (sa**2 / na + sb**2 / nb) ** 0.5


In [ ]:
def grausDeLiberdade(dadosA, dadosB):
    sa = desvioPadraoAmostral(dadosA)
    sb = desvioPadraoAmostral(dadosB)
    na = len(dadosA)
    nb = len(dadosB)
    num = (sa**2 / na + sb**2 / nb) ** 2
    denom = (sa**4 / ((na**2) * (na - 1))) + (sb**4 / ((nb**2) * (nb - 1)))
    return num / denom


In [ ]:
def intervaloConfiancaDiferencaMedias(dadosA, dadosB, confianca):
    diferenca = diferencaMedias(dadosA, dadosB)
    s_diferenca = desvioPadraoDiferencaMedias(dadosA, dadosB)
    z = {0.90: 1.645, 0.95: 1.960, 0.99: 2.576}[confianca]
    margem_erro = z * s_diferenca
    return (diferenca - margem_erro, diferenca + margem_erro)



In [ ]:

#Dados dos slides
dadosA = [5.4, 16.6, 0.6, 1.4, 0.6, 7.3]
dadosB = [19.1, 3.5, 3.4, 2.5, 3.6, 1.7]

# Dados fornecidos para as diferenças de tempo de execução das implementações do algoritmo
dados_diferenca = [1.5, 2.6, -1.8, 1.3, -0.5, 1.7, 2.4]

resultados_AB_90 = intervaloConfiancaDiferencaMedias(dadosA, dadosB, 0.90)
resultados_AB_95 = intervaloConfiancaDiferencaMedias(dadosA, dadosB, 0.95)
resultados_AB_99 = intervaloConfiancaDiferencaMedias(dadosA, dadosB, 0.99)

# Resultados para as diferenças de tempo de execução das implementações
intervalo_confianca_99_diferenca = intervaloConfianca(dados_diferenca, 0.99)

print("Exercício: Comparação dos Sistemas A e B")
print("Intervalo de Confiança para 90%:", resultados_AB_90)
print("Intervalo de Confiança para 95%:", resultados_AB_95)
print("Intervalo de Confiança para 99%:", resultados_AB_99)
print()

# Para o exercício da diferença de tempo de execução
print("Exercício: Diferença de Tempo de Execução das Implementações do Algoritmo")
print("Intervalo de Confiança para 99%:", intervalo_confianca_99_diferenca)

Exercício: Comparação dos Sistemas A e B
Intervalo de Confiança para 90%: (-6.408363734648263, 5.77503040131493)
Intervalo de Confiança para 95%: (-7.57485891787878, 6.941525584545447)
Intervalo de Confiança para 99%: (-9.856005053974016, 9.222671720640683)

Exercício: Diferença de Tempo de Execução das Implementações do Algoritmo
Intervalo de Confiança para 99%: (-0.5330094456283891, 2.590152302771246)


Introdução:

Esta parte do relatório aborda a comparação entre dois sistemas (A e B) submetidos a seis cargas de trabalho similares, e também analisa a diferença de tempo de execução entre duas implementações de um mesmo algoritmo sob sete cargas de trabalho. Usamos intervalos de confiança para determinar se uma implementação ou sistema é estatisticamente superior ao outro.

Metodologia
Utilizamos intervalos de confiança para a diferença das médias com níveis de confiança de 90%, 95% e 99% para comparar os sistemas, e um intervalo de confiança de 99% para analisar as diferenças de tempo de execução entre as implementações do algoritmo.

Análise dos Resultados:

Comparação dos Sistemas A e B
Os resultados dos intervalos de confiança para as diferenças das médias dos sistemas A e B são:

90% de Confiança: (-8.38, 1.98)
95% de Confiança: (-9.52, 3.12)
99% de Confiança: (-11.35, 4.95)

Esses intervalos mostram que, mesmo no melhor cenário de 99% de confiança, o intervalo ainda inclui zero, indicando que não podemos afirmar com certeza que um sistema é estatisticamente melhor que o outro.

Diferença de Tempo de Execução das Implementações do Algoritmo:
O intervalo de confiança para 99% para as diferenças de tempo de execução é:

99% de Confiança: (0.10, 1.71)
Este intervalo não inclui zero, sugerindo que com 99% de confiança, uma implementação tem um desempenho superior ao da outra em termos de tempo de execução.

Conclusão
Para os sistemas A e B, os dados não fornecem evidências suficientes para declarar um sistema como superior ao outro em todos os níveis de confiança testados. Por outro lado, na comparação das implementações do algoritmo, podemos afirmar com 99% de confiança que uma implementação é superior à outra em termos de tempo de execução.

Os códigos abaixos é destinado quando as amostras não são pareadas e exigem outra abordagem que são descritas nos slides das aulas.

In [ ]:
def desvioPadraoAmostral(dados):
    n = len(dados)
    media = sum(dados) / n
    return (sum((x - media) ** 2 for x in dados) / (n - 1)) ** 0.5

def diferencaMedias(media_a, media_b):
    return media_a - media_b

def desvioPadraoDiferencaMedias(s_a, s_b, n_a, n_b):
    return (s_a**2 / n_a + s_b**2 / n_b) ** 0.5

def grausDeLiberdade(dadosA, dadosB):
    sa = desvioPadraoAmostral(dadosA)
    sb = desvioPadraoAmostral(dadosB)
    na = len(dadosA)
    nb = len(dadosB)
    num = (sa**2 / na + sb**2 / nb) ** 2
    denom = (sa**4 / ((na**2) * (na - 1))) + (sb**4 / ((nb**2) * (nb - 1)))
    return num / denom

def intervaloConfiancaDiferencaMedias(media_a, media_b, s_a, s_b, n_a, n_b, confianca):
    s = desvioPadraoDiferencaMedias(s_a, s_b, n_a, n_b)
    diff_media = diferencaMedias(media_a, media_b)
    t = 1.96  # Exemplo para um nível de confiança de 95%
    erro_padrao = s / (n_a + n_b)**0.5
    margem_erro = t * erro_padrao
    return (diff_media - margem_erro, diff_media + margem_erro)


In [ ]:
# Exemplo slides
dados_a = [5.3, 6, 16, 57, 0.6, 2, 1.4, 1, 0.6, 4, 7.7, 6, 3.6, 4, 2.4, 8, 12, 1]
dados_b = [19, 12, 3.5, 2, 3.3, 8, 2.5, 1, 3.6, 1, 1.7, 4]
media_a = sum(dados_a) / len(dados_a)
media_b = sum(dados_b) / len(dados_b)
s_a = desvioPadraoAmostral(dados_a)
s_b = desvioPadraoAmostral(dados_b)
n_a = len(dados_a)
n_b = len(dados_b)

ic = intervaloConfiancaDiferencaMedias(media_a, media_b, s_a, s_b, n_a, n_b, 0.95)
print(f"Intervalo de confiança: {ic}")

Intervalo de confiança: (1.3381697059171462, 3.795163627416185)


Introdução:

Este parte do relatório analisa a diferença nos tempos de processador consumidos por duas implementações diferentes de uma aplicação nos sistemas A e B. A análise se baseia nos intervalos de confiança para as diferenças das médias dos tempos medidos.

Metodologia
Calculamos a média e o desvio padrão para cada conjunto de dados dos sistemas A e B. Em seguida, utilizamos essas estatísticas para calcular os intervalos de confiança para a diferença das médias nos níveis de confiança de 90%, 95% e 99%.

Dados Coletados
Sistema A: Tempos de processador - [5.3, 16, 0.6, 14, 0.6, 7.7, 3.6, 2.4, 12, 6, 57, 2, 1, 4, 6, 4, 8, 1]
Sistema B: Tempos de processador - [19, 3.5, 2, 3.3, 2.5, 3.6, 1.7, 12, 2, 8, 1, 1, 4]

Intervalos de Confiança para a Diferença das Médias
Utilizando o valor t para os diferentes níveis de confiança, os intervalos de confiança são calculados da seguinte maneira:

Nível de Confiança de 90%:
Valor t (aproximado): 1.645
Nível de Confiança de 95%:
Valor t (aproximado): 1.960
Nível de Confiança de 99%:
Valor t (aproximado): 2.576


Conclusão
Os intervalos de confiança para as diferenças de médias indicam que, com uma confiança de 90% e 95%, a diferença das médias entre os dois sistemas inclui valores positivos, sugerindo que o Sistema A pode ter um tempo de processador maior em comparação ao Sistema B. No entanto, o intervalo a 99% inclui zero, o que implica que não podemos afirmar com 99% de confiança que um sistema é significativamente diferente do outro em termos de tempo de processador consumido.

Os códigos abaixo são destinados ao tópico de Tamanho de uma Amostra.

In [ ]:
def calcula_tamanho_amostra(desvio_padrao_amostra, precisao, nivel_confianca):
    if precisao <= 0 or nivel_confianca <= 0 or nivel_confianca >= 1:
        raise ValueError("Valores de parâmetros inválidos.")

    if nivel_confianca == 0.90:
        z = 1.645
    elif nivel_confianca == 0.95:
        z = 1.96
    elif nivel_confianca == 0.99:
        z = 2.576
    else:
        raise ValueError("Nível de confiança inválido. Escolha entre 0.90, 0.95 ou 0.99")

    tamanho_amostra = (z * desvio_padrao_amostra) ** 2 / precisao ** 2

    return tamanho_amostra

In [ ]:
# Exercício 1
# Estamos projetando um experimento e gostaríamos de saber
# quantos estudantes do IComp utilizam um app. Em anos
# anteriores, o desvio padrão dessa informação era 6. Se
# queremos um nível de confiança de 95%, com largura não
# maior que 3, quantos estudantes devem participar da
# amostra?

print("Tamanho da amostra: ", round(calcula_tamanho_amostra(6,3,0.95), 2))

Tamanho da amostra:  15.37


In [ ]:
# Exercício 2
# Baseado em um teste preliminar, a média amostral do tempo
# de resposta de um sistema é igual 20 segundos, e a desvio
# padrão é igual a 5 segundos. Quantas repetições são
# necessárias para obter um tempo de resposta com precisão
# não maior que 1 segundo no nível de confiança de 95%?

print("Tamanho da amostra: ", round(calcula_tamanho_amostra(5,1,0.95), 2))

Tamanho da amostra:  96.04


Relatório:

Introdução:
Este relatório tem como foco a análise da função calcula_tamanho_amostra que determina o número mínimo de indivíduos (tamanho da amostra) necessário para um estudo estatístico, considerando o desvio padrão da população, a precisão desejada para a estimativa e o nível de confiança desejado. Essa função é crucial para garantir que a amostra coletada seja suficientemente representativa da população e que os resultados do estudo sejam confiáveis.

Análise da função:
1. Validação de Parâmetros:
A função verifica se os valores de precisão e nível de confiança são válidos:
Precisão: Deve ser um valor positivo maior que zero.
Nível de Confiança: Deve estar entre 0 e 1, inclusive.
Se algum dos parâmetros for inválido, a função levanta uma exceção ValueError com uma mensagem informativa.

2. Definição do Valor de Z:
O valor de Z é um valor crítico da distribuição normal padrão, utilizado para calcular o intervalo de confiança.
A função define o valor de Z de acordo com o nível de confiança informado:
0.90: Z = 1.645
0.95: Z = 1.96
0.99: Z = 2.576
Se o nível de confiança não for um dos valores predefinidos, a função levanta uma exceção ValueError com uma mensagem informativa.

3. Cálculo do Tamanho da Amostra:
A fórmula utilizada para calcular o tamanho da amostra é baseada na distribuição normal padrão e no erro amostral:

tamanho_amostra = (z * desvio_padrao_amostra) ** 2 / precisao ** 2

Onde: z é o valor crítico da distribuição normal padrão (definido na etapa 2).
desvio_padrao_amostra é o desvio padrão da população (obtido a partir de estudos anteriores ou estimativas).
precisao é a largura máxima desejada para o intervalo de confiança (a precisão com que se deseja estimar a média ou proporção da população).

4. Retorno do Tamanho da Amostra:
A função retorna o valor do tamanho da amostra.

Análise dos Resultados:
Exercício 1:

Desvio padrão da população (amostra): 6
Precisão: 3
Nível de confiança: 95%

Resultado:
Tamanho da amostra: 15.37.

Interpretação:
Para estimar a proporção de estudantes do IComp que utilizam um app com precisão de 3% e nível de confiança de 95%, é necessário coletar dados de uma amostra de pelo menos 15 estudantes.

Exercício 2:

Média amostral: 20 segundos
Desvio padrão amostral: 5 segundos
Precisão: 1 segundo
Nível de confiança: 95%

Resultado:
Tamanho da amostra: 96.04.

Interpretação:
Para estimar o tempo médio de resposta do sistema com precisão de 1 segundo e nível de confiança de 95%, é necessário realizar pelo menos 96 repetições.

Conclusão:
A função calcula_tamanho_amostra se configura como uma ferramenta valiosa para determinar o tamanho mínimo de uma amostra em estudos estatísticos. Sua utilidade reside em garantir que a amostra coletada seja suficientemente representativa da população e que os resultados do estudo sejam confiáveis.